### PR TITLE
feat(mac): mnemonic layout support 🐘 

### LINKED ISSUE #7068: feat(windows): mcompile - use static tables for Windows keyboard data
This is a proposal to decouple the Windows system keyboard data from the kbdxx.dll file so that we can use mcompile on other platforms as well. The idea would be to have a program to query the set of Windows kbdxx DLLs that we are interested in, and write each Windows keyboard's data to a datafile. Then mcompile would load this datafile rather than the Windows DLL when running the mcompile process.

Sample call currently:

```
mcompile -d c:\projects\keyman\keyboards\release\sil\sil_ipa\build\sil_ipa.kmx kbdgr.dll 0407 c:\projects\keyman\keyboards\release\sil\sil_ipa\build\sil_ipa_german.kmx
```

mcompile should add two new command line modes:

First, exporting keyboard dll data to file:
```
mcompile -e kbdgr.dll 0407 kbdgr.txt
```

Second, using keyboard dll data from file:
```
mcompile -d c:\projects\keyman\keyboards\release\sil\sil_ipa\build\sil_ipa.kmx kbdgr.txt c:\projects\keyman\keyboards\release\sil\sil_ipa\build\sil_ipa_german.kmx
```

The second part of this project would be to remove Windows-isms from mcompile so that we can build it on any desktop platform.

### PR DESCRIPTION
**Pull request by SabineSIL**
fixes #7068  
see #3345
 see #9384

This is a proposal to rewrite mcompile for macOS.

It will be a standalone implementation for macOS for now which is similar to the standalone implementation for Linux. (#9384)

To achieve this we can reuse most of the code from mcompile-Linux, especially the structure and the .kmx rewriting. We need to adapt the keyboard interrogation though. On Linux we use x11; For macOS we use the Carbon.h library.

Since we keep the structure of the code as it is on Linux, we only have to change the content of several functions, especially those that handle the keyboard interrogation and the mapping between Keycodes/Scancodes and Keyvalues.

Therefore for mcompile-mac we will first set up the the .kmx rewriting, then add the handling of non-deadkeys and in the last step add the handling of deadkeys. 

Maybe in a next step we may combine the 3 versions of mcompile for Windows, Linux, macOS into a cross-platform version of mcompile.

 @keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
Test - ... macOS build failures appear to be due to the age of the PR's branch:

> This branch is [28 commits ahead of](https://github.com/keymanapp/keyman/compare/master...feat/mac/mcompile_mac), [9579 commits behind](https://github.com/keymanapp/keyman/compare/feat/mac/mcompile_mac...master) master.
[#11334](https://github.com/keymanapp/keyman/pull/11334)

From the Test (Keyman - macOS) build log:

```
16:32:43   /Users/marc_durdin/buildAgent/work/99b311828f4ee7c/keyman/mac/KeymanEngine4Mac/KeymanEngine4Mac.xcodeproj: warning: The macOS deployment target 'MACOSX_DEPLOYMENT_TARGET' is set to 10.10, but the range of supported deployment target versions is 10.13 to 14.4.99. (in target 'KeymanEngine4Mac' from project 'KeymanEngine4Mac')
16:32:43   warning: Run script build phase 'Assert version number same as git branch' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase. (in target 'KeymanEngine4Mac' from project 'KeymanEngine4Mac')
16:32:43   warning: Run script build phase 'Run Script' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase. (in target 'KeymanEngine4Mac' from project 'KeymanEngine4Mac')
16:32:43   /Users/marc_durdin/buildAgent/work/99b311828f4ee7c/keyman/mac/KeymanEngine4Mac/KeymanEngine4Mac.xcodeproj: warning: The macOS deployment target 'MACOSX_DEPLOYMENT_TARGET' is set to 10.10, but the range of supported deployment target versions is 10.13 to 14.4.99. (in target 'KeymanEngine4MacTests' from project 'KeymanEngine4Mac')
```

Excerpt from _that_:

> warning: The macOS deployment target 'MACOSX_DEPLOYMENT_TARGET' is set to 10.10, but the range of supported deployment target versions is 10.13 to 14.4.99.

This is a setting that would have been updated when we did the Xcode 15 update in #11302.

I'm going to go ahead and update the base-branch via "Update branch" merge.

**Comment by mcdurdin** [comment]
There is an issue somewhere here with c++ `-O1`, `-O2`, `-O3`. We have found that unless we use `-O0` the results are invalid. This suggests that somewhere in our code we have something that is not quite C++ standards-compliant, and we should try and track that down. For now, we will use `-O0` so we can move forward.

The exact commands we have been using for the test:

```
/usr/bin/c++ -O1 -fdiagnostics-color=always -g -framework Carbon /Users/sabine/projects/keyman/keyman/mac/mcompile/*.cpp /Users/sabine/projects/keyman/keyman/common/cpp/km_u16.cpp  -o /Users/sabine/projects/keyman/keyman/mac/mcompile/executable/mcompile-bad.exe -std=c++17 -I /Users/sabine/projects/keyman/keyman/common/include/
```

```
./executable/mcompile-bad.exe test_mcompile_mac_01.kmx res_bad.kmx
```

This results in:

```
Add deadkey rule:  + [0 K_192] > dk(1)
Add deadkey rule:  + [0 K_187] > dk(2)
Add deadkey rule:  + [16 K_187] > dk(2)
Did not find a match for mnemonic rule on line 14, + '`' > ...
```

If we use `-O0` in the c++ compile, we instead get the correct result (note `dk(3)` vs `dk(2)` on line 3 as well as the error):

```
Add deadkey rule:  + [0 K_192] > dk(1)
Add deadkey rule:  + [0 K_187] > dk(2)
Add deadkey rule:  + [16 K_187] > dk(3)
```

The source keyboard:

```
c test_mcompile_mac_01 generated from template at 2024-08-15 08:36:57
c with name "test_mcompile_mac_01"
store(&VERSION) '10.0'
store(&NAME) 'test_mcompile_mac_01'
store(&COPYRIGHT) 'Copyright ©'
store(&KEYBOARDVERSION) '1.0'
store(&TARGETS) 'windows macosx linux'
store(&mnemoniclayout) '1'

begin Unicode > use(main)

group(main) using keys

+ '`' > 'test'
```

File: mac/mcompile/keymap.h [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+  ShftCtrl     = Shft | Ctrl,         // 3
+  Menu         = 4,                   // 4 -- NOT USED
+  ShftMenu     = Shft | Menu,         // 5 -- NOT USED
+  MenuCtrl     = Menu | Ctrl,         // 6
+  ShftMenuCtrl = Shft | Menu | Ctrl,  // 7
+  Xxxx         = 8,                   // 8
+  ShftXxxx     = Shft | Xxxx,         // 9
+};
+
+#define VK_SPACE    0x20
```
Comment (mcdurdin) [inline mac/mcompile/keymap.h:25]: These are another candidate for cleanup in various source files (should be a separate PR on its own)

File: mac/mcompile/keymap.h [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+#define VK_LMENU          0xA4
+#define VK_RMENU          0xA5
+
+#define VK_SHIFT          0x10
+#define VK_CONTROL        0x11
+#define VK_MENU           0x12
+#define VK_PAUSE          0x13
+#define VK_CAPITAL        0x14
+// Map of all US English virtual key codes that we can translate
+const KMX_DWORD KMX_VKMap[] = {
```
Comment (mcdurdin) [inline mac/mcompile/keymap.h:58]: This should be defined in the .cpp and only declared as an extern here:

```suggestion
extern const KMX_DWORD KMX_VKMap[];
```

File: mac/mcompile/keymap.h [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+static KMX_DWORD keycode_max  = 50;
+static int keycode_spacebar   = 49;
+static KMX_DWORD max_shiftstate = 10;
+
+// shiftstates we can use for mac
+static int MAC_BASE = 0;
+static int MAC_SHIFT = 2;
+static int MAC_OPT = 8;
+static int MAC_SHIFT_OPT = 10;
+static int ss_mac[] = {MAC_BASE, MAC_SHIFT, MAC_OPT, MAC_SHIFT_OPT};
```
Comment (mcdurdin) [inline mac/mcompile/keymap.h:98]: These should all be externs in the .h and defined in the .cpp

File: mac/mcompile/keymap.h [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+int mac_append_underlying_ToVector(vec_dword_3D& all_vector, const UCKeyboardLayout* keyboard_layout);
+
+/** @brief create a pointer to pointer of the current keymap for later use */
+bool mac_InitializeUCHR(const UCKeyboardLayout** keyboard_layout);
+
+/**
+ * @brief  array of USVirtualKey-ScanCode-pairs
+ *         we use the same type of array as throughout Keyman even though we have lots of unused fields
+ */
+const KMX_DWORD mac_USVirtualKeyToScanCode[256] = {
```
Comment (mcdurdin) [inline mac/mcompile/keymap.h:128]: ditto

File: mac/mcompile/keymap.h [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+	0xFFFF, 		    // not used
+	0xFFFF, 		    // not used
+	0xFFFF, 		    // not used
+};
+
+/**
+ * @brief  array of ScanCode-USVirtualKey-pairs
+ * 				 we use the same type of array as throughout Keyman even though we have lots of unused fields
+ */
+const KMX_DWORD mac_ScanCodeToUSVirtualKey[128] = {
```
Comment (mcdurdin) [inline mac/mcompile/keymap.h:391]: ditto

File: mac/mcompile/keymap.cpp [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+ * @brief  write data of the US keyboard into a 3D-Vector which later will contain
+ *         data of the US keyboard and the currently used (underlying) keyboard
+ * @param[in,out] vec_us Vector that holds the data of the US keyboard
+ * @return 0 on success;
+ * 				 1 if data of US keyboard was not written;
+ */
+int mac_write_US_ToVector(vec_dword_3D& vec_us) {
+  // Values for US        :     A,  S,  D,  F,  H,  G,  Z,  X, C,  V,  §, B,  Q,  W,  E,  R,  Y,  T, 1, 2, 3, 4, 6, 5, =, 9, 7, -, 8, 0,  ],  O,  U,  [,  I,  P,CR,  L,  J, ',  K, ;,  \, ,, /,  N,  M, .
+  std::vector<int> us_Base  = {97,115,100,102,104,103,122,120,99,118,167,98,113,119,101,114,121,116,49,50,51,52,54,53,61,57,55,45,56,48, 93,111,117, 91,105,112,13,108,106,39,107,59, 92,44,47,110,109,46};
+  std::vector<int> us_Shift = {65, 83, 68, 70, 72, 71, 90, 88,67, 86,177,66, 81, 87, 69, 82, 89, 84,33,64,35,36,94,37,43,40,38,95,42,41,125, 79, 85,123, 73, 80,13, 76, 74,34, 75,58,124,60,63, 78, 77,62};
```
Comment (mcdurdin) [inline mac/mcompile/keymap.cpp:108]: It'd be good for these constants to be defined outside the function (and as consts)

**Review by mcdurdin — CHANGES_REQUESTED** [review]
This is looking good. A number of changes requested!

**Comment by SabineSIL** [comment]
As a solution to the above cited problem [ with c++ -O1, -O2, -O3. ](https://github.com/keymanapp/keyman/pull/11334#issuecomment-2290784399) we found 2 things:

1. in  mc_kmxfile.cpp - KMX_CopyKeyboard() add:
  
   ```
   kbp->dpBitmapOffset = ckbp->dpBitmapOffset;
   kbp->dwBitmapSize = ckbp->dwBitmapSize;
   ```
  

2. in keymap.cpp - mac_KMX_get_KeyValUnderlying_From_KeyCodeUnderlying change 

   
   ` dky = (PKMX_WCHAR)(std::u16string(1, keyV)).c_str();`

     to
   

     ```
    PKMX_WCHAR dky = NULL;
    std::u16string keyVS(1, keyV);
    dky = (PKMX_WCHAR) keyVS.c_str();
    ```

File: mac/mcompile/mcompile.cpp [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+ * @param  kbd        pointer to the keyboard
+ * @param  vk_US      virtual key of the us keyboard
+ * @param  shift      shiftstate
+ * @param  deadkey    character produced by a deadkey
+ * @param  all_vector vector that holds the data of the US keyboard and the currently used (underlying) keyboard
+ * @param  keymap     pointer to the currently used (underlying) keyboard Layout
+ * @param  dk_Table   a vector of all possible deadkey combinations for all Linux keyboards
+ */
+void mac_KMX_ConvertDeadkey(LPKMX_KEYBOARD kbd, KMX_WORD vk_US, KMX_DWORD shift, KMX_WCHAR deadkey, vec_dword_3D& all_vector, const UCKeyboardLayout* keyboard_layout) {
+  KMX_WORD deadkeys[512] = {0};
```
[the remark this answers is not in the document]
Comment (SabineSIL) [reply mac/mcompile/mcompile.cpp]: - changed to 1000 dk-pairs (=3000+3 elements in dk array) Do we need more?
- in case of overflow: 
  - stop processing more dk and 
  - write out a warning ("x deadkeys have not been processed")

File: mac/mcompile/mcompile.cpp [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+
+  #elif ((__linux__) || (__unix__))       // LINUX, UNIX
+    if (mac_KMX_DoConvert(kmxfile, bDeadkeyConversion, argc)) {  // I4552F
+      KMX_SaveKeyboard(kmxfile, outfile);
+  }
+
+  #elif (__APPLE__ && TARGET_OS_MAC)      // macOS
+    if (mac_KMX_DoConvert(kmxfile, bDeadkeyConversion, argc)) {  // I4552F
+      KMX_SaveKeyboard(kmxfile, outfile);
+  }
```
[the remark this answers is not in the document]
Comment (SabineSIL) [reply mac/mcompile/mcompile.cpp]: On Windows-mcompile nothing happens. What do we want to happen?

File: mac/mcompile/mcompile.cpp [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+ * @param  shift   shiftstate
+ * @param  ch      character of the underlying keyboard
+ */
+void mac_KMX_TranslateDeadkeyKey(LPKMX_KEY key, KMX_WCHAR deadkey, KMX_WORD vk, KMX_DWORD shift, KMX_WORD ch) {
+   if ((key->ShiftFlags == 0 || key->ShiftFlags & VIRTUALCHARKEY) && key->Key == ch) {
+    // The weird LCTRL+RALT is Windows' way of mapping the AltGr key.
+    // We store that as just RALT, and use the option "Simulate RAlt with Ctrl+Alt"
+    // to provide an alternate..
+  if ((shift & (LCTRLFLAG | RALTFLAG)) == (LCTRLFLAG | RALTFLAG))  // I4327
+      shift &= ~LCTRLFLAG;
```
[the remark this answers is not in the document]
Comment (SabineSIL) [reply mac/mcompile/mcompile.cpp]: **bu_phonetic_WIN_DE_w67.kmn:**
    bu_phonetic.kmx on mcompile-windows; language German; with alt+shift/alt keys processed
**bu_phonetic_out_DE_w67.kmn:**  
   bu_phonetic.kmx on mcompile-mac; language German; with alt+shift/alt keys processed
**std_mac_bu_phonetic_out_DE_w67.kmn:**  
   bu_phonetic.kmx on mcompile-mac; language German; with alt+shift/alt keys processed ;standard - above flags used
Comparison

File: mac/mcompile/keymap.cpp [hunk 10]
```diff
@@ -0,0 +1,913 @@
...
+  unicodeString[0] = 0;
+
+  if (!ensureValidInputForKeyboardTranslation(shiftstate_mac, keycode))
+    return 0;
+
+  /*
+    UCKeyTranslate != 0 if a dk was found; then run UCKeyTranslate again with a SPACE (keycode_spacebar) to get the plain dk e.g.'^'.
+    If CAPS is used: always add 4 e.g. SHIFT = 2; SHIFT+CAPS = 6
+  */
+  status = UCKeyTranslate(keyboard_layout, keycode, kUCKeyActionDown, (shiftstate_mac + 4 * caps), LMGetKbdType(), keyTranslateOptions, &deadkeystate, maxStringlength, &actualStringlength, unicodeString);
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/mcompile/keymap.cpp:676]: From the description in the commented code, it looks like there is a memory corruption problem.

File: mac/mcompile/mcompile.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/mcompile/mcompile.cpp]: Better to dynamically allocate I think. Just use a `std::vector<KMX_WORD>` and dynamically expand

File: mac/mcompile/mcompile.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/mcompile/mcompile.cpp]: I see -- looks like we do need to maintain this code then

File: mac/mcompile/mcompile.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/mcompile/mcompile.cpp]: It should return an error but then we need a way to report it ...

File: mac/mcompile/keymap.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (SabineSIL) [reply mac/mcompile/keymap.cpp:676]: resolved it: 
check status of UCKeyTranslate  `if(status != noErr)`  can only be done after the _second_ use of UCKeyTranslate() and not between the two uses of that function.

File: mac/mcompile/mcompile.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (SabineSIL) [reply mac/mcompile/mcompile.cpp]: OK changed

File: mac/mcompile/mcompile.cpp [hunk 11]
```diff
@@ -0,0 +1,597 @@
...
+  for (unsigned int i = 0; i < kbd->cxGroupArray; i++) {
+    if (kbd->dpGroupArray[i].fUsingKeys) {
+      LPKMX_KEY keys = new KMX_KEY[kbd->dpGroupArray[i].cxKeyArray + 1];
+      memcpy(keys + 1, kbd->dpGroupArray[i].dpKeyArray, kbd->dpGroupArray[i].cxKeyArray * sizeof(KMX_KEY));
+      keys[0].dpContext = new KMX_WCHAR[1];
+      keys[0].dpContext[0] = 0;
+      keys[0].dpOutput = new KMX_WCHAR[4];
+      keys[0].dpOutput[0] = UC_SENTINEL;
+      keys[0].dpOutput[1] = CODE_DEADKEY;
+      keys[0].dpOutput[2] = deadkey; // TODO: translate to unique index
```
Comment (mcdurdin) [inline mac/mcompile/mcompile.cpp:302]: Is this TODO resolved?

**Review by mcdurdin — APPROVED** [review]
LGTM. Let's get this in and finish off this epic. I know we'll still need to do more work to integrate into the core products but at least this will be ready to go

File: mac/mcompile/mcompile.cpp (hunk 11 above)
Comment (SabineSIL) [reply mac/mcompile/mcompile.cpp:302]: this is a comment from mcompile-windows - I did not do any changes
